db/db_impl/db_impl.h

```diff
@@ -1046,6 +1046,52 @@ class DBImpl : public DB {
   // flush LOG out of application buffer
   void FlushInfoLog();
 
+  // Interface to block and signal the DB in case of stalling writes by
+  // WriteBufferManager. Each DBImpl object contains ptr to WBMStallInterface.
+  // When DB needs to be blocked or signalled by WriteBufferManager,
+  // state_ is changed accordingly.
+  class WBMStallInterface : public StallInterface {
+   public:
+    enum State {
+      BLOCKED = 0,
+      RUNNING,
+    };
+
+    WBMStallInterface() : state_cv_(&state_mutex_) {
+      MutexLock lock(&state_mutex_);
+      state_ = State::RUNNING;
+    }
+
+    // Change the state_ to State::BLOCKED and wait until its state is
+    // changed by WriteBufferManager. When stall is cleared, Signal() is
+    // called to change the state and unblock the DB.
+    void Block() override {
+      MutexLock lock(&state_mutex_);
+      state_ = State::BLOCKED;
```

Comment:
I'm kind of concerned about missed signals. I think a better sequence would be as follows -
```
state_ = State::BLOCKED;
wbm->BeginWriteStall(this);
MutexLock lock(&state_mutex_);
while (state_ == State::BLOCKED) {
  state_cv_.Wait();
}
```

Reply:
But lock protects ```state_``` so shouldn't ```state_ = State::BLOCKED;``` be inside the lock? Also can you explain about missed signals, I didn't get it.

Reply:
The current sequence is roughly as follows -
```
wbm->BeginWriteStall(wbm_stall_);
mutex_.Unlock();
...
wbm_stall_->Block();
```
The problem is, between the first ```mutex_.Unlock()``` and the call to ```Block()```, WBM could have signalled this thread. That signal will be lost, and ```Block()``` will call ```state_cv_.Wait()``` and not get woken up.